packages/docs-ui/src/services/doc-event-manager.service.ts

```diff
@@ -0,0 +1,322 @@
+/**
+ * Copyright 2023-present DreamNum Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { DocumentDataModel, ICustomRange, IParagraph, ITextRangeParam, Nullable } from '@univerjs/core';
+import { Disposable, fromEventSubject, Inject } from '@univerjs/core';
+import { DocSkeletonManagerService } from '@univerjs/docs';
+import type { Documents, DocumentSkeleton, IBoundRectNoAngle, IMouseEvent, IPointerEvent, IRender, IRenderContext, IRenderModule } from '@univerjs/engine-render';
+import { getLineBounding, NodePositionConvertToCursor, pxToNum, TRANSFORM_CHANGE_OBSERVABLE_TYPE } from '@univerjs/engine-render';
+import { distinctUntilChanged, filter, Subject, throttleTime } from 'rxjs';
+import { transformOffset2Bound } from './doc-popup-manager.service';
+
+interface ICustomRangeBound {
+    customRange: ICustomRange;
+    rects: IBoundRectNoAngle[];
+    segmentId?: string;
+    segmentPageIndex: number;
+}
+
+interface IBulletBound {
+    rect: IBoundRectNoAngle;
+    segmentId?: string;
+    segmentPageIndex: number;
+    paragraph: IParagraph;
+}
+
+const calcDocRangePositions = (range: ITextRangeParam, documents: Documents, skeleton: DocumentSkeleton, pageIndex: number): IBoundRectNoAngle[] | undefined => {
+    const startPosition = skeleton.findNodePositionByCharIndex(range.startOffset, true, range.segmentId, pageIndex);
+    const endPosition = skeleton.findNodePositionByCharIndex(range.endOffset, true, range.segmentId, pageIndex);
+    // console.log('===startPosition', range, startPosition, endPosition);
+    if (!endPosition || !startPosition) {
+        return;
+    }
+
+    const documentOffsetConfig = documents.getOffsetConfig();
+    const convertor = new NodePositionConvertToCursor(documentOffsetConfig, skeleton);
+    const { borderBoxPointGroup } = convertor.getRangePointData(startPosition, endPosition);
+    const bounds = getLineBounding(borderBoxPointGroup);
+
+    return bounds;
+};
+
+interface ICustomRangeActive {
+    range: ICustomRange; segmentId?: string; segmentPageIndex: number;
+}
+
+interface IBulletActive {
+    paragraph: IParagraph; segmentId?: string; segmentPageIndex: number;
+}
+
+export class DocEventManagerService extends Disposable implements IRenderModule {
+    private _hoverCustomRanges$ = new Subject<ICustomRangeActive[]>();
```

Comment:
```suggestion
    private readonly _hoverCustomRanges$ = new Subject<ICustomRangeActive[]>();
```

Use `readonly` if possible.